diff --git a/.bazelversion b/.bazelversion
index 8104cab..acd405b 100644
--- a/.bazelversion
+++ b/.bazelversion
@@ -1 +1 @@
-8.1.0
+8.6.0
diff --git a/build/buildkite/pipeline.yml b/build/buildkite/pipeline.yml
index eed342e..08d703d 100644
--- a/build/buildkite/pipeline.yml
+++ b/build/buildkite/pipeline.yml
@@ -5,7 +5,7 @@ env:
 _templates:
   - &init . ${HOME}/init.sh && bazel version
   - &k8s-buildkite-plugin-base
-    image: "us-central1-docker.pkg.dev/buildkite-382710/buildkite-production-docker/buildkite-bazel@sha256:3a858aeb09abeb19f41f888a7471e8869229d7b231370eb3d3793794863928a0"
+    image: "us-central1-docker.pkg.dev/buildkite-382710/buildkite-production-docker/buildkite-bazel@sha256:503e0797a9f8a2a96ca478b8239043b8d980bf0ab94136f5242429b2af950bf4"
     secret-name: "buildkite"
     environment-from-secret: "buildkite"
     agent-token-secret-key: "buildkite-agent-token"
diff --git a/MODULE.bazel b/MODULE.bazel
index 20337db..a4db5b1 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -6,10 +6,9 @@ module(
 bazel_dep(name = "rules_go", version = "0.59.0", dev_dependency = True)
 bazel_dep(name = "gazelle", version = "0.47.0", dev_dependency = True)
 bazel_dep(name = "rules_oci", version = "2.2.6", dev_dependency = True)
-bazel_dep(name = "aspect_bazel_lib", version = "2.21.2", dev_dependency = True)
-bazel_dep(name = "rules_python", version = "1.7.0", dev_dependency = True)  # TODO: remove after user root issue is released and rules pkg bumped: https://github.com/bazelbuild/rules_python/pull/713 (rules_pkg)
-bazel_dep(name = "bazel_skylib", version = "1.8.2", dev_dependency = True)
-bazel_dep(name = "rules_pkg", version = "1.1.0", dev_dependency = True)
+bazel_dep(name = "aspect_bazel_lib", version = "2.22.5", dev_dependency = True)
+bazel_dep(name = "bazel_skylib", version = "1.9.0", dev_dependency = True)
+bazel_dep(name = "rules_pkg", version = "1.2.0", dev_dependency = True)
 bazel_dep(name = "container_structure_test", version = "1.21.1", dev_dependency = True)
 bazel_dep(name = "aspect_rules_lint", version = "1.10.2", dev_dependency = True)
 bazel_dep(name = "buildifier_prebuilt", version = "8.2.1", dev_dependency = True)
diff --git a/BUILD.bazel b/BUILD.bazel
index a2009b8..308a2b9 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -1,4 +1,5 @@
 load("@gazelle//:def.bzl", "gazelle")
+load("@rules//pkg/ai/llm:llm.bzl", "earth_go_service_agents_md_template")
 load("@rules//pkg/ls_lint:ls_lint.bzl", "ls_lint_test")
 load("@rules_go//go:def.bzl", "TOOLS_NOGO", "nogo")
 
@@ -44,3 +45,11 @@ ls_lint_test(
         "//deployments/staging:files",
     ],
 )
+
+earth_go_service_agents_md_template(
+    name = "agents_md",
+    service = [
+        "email",
+    ],
+    db = "mysql",
+    visibility = ["//visibility:private"],
+)
diff --git a/.bazelrc b/.bazelrc
index dca8d94..1ca6e5f 100644
--- a/.bazelrc
+++ b/.bazelrc
@@ -1 +1,9 @@
-build --define=MINDFUL_USER= --workspace_status_command="bash scripts/bazel/workspace_status.sh"
+build --define=MINDFUL_USER=
+build --workspace_status_command="bash scripts/bazel/workspace_status.sh"
+
+# prebuilt protobuf
+# remove after bazel 9 migration and https://github.com/bazel-contrib/rules_go/issues/4587
+build --incompatible_enable_proto_toolchain_resolution
+build --per_file_copt=external/.*protobuf.*@-DPROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
+build --host_per_file_copt=external/.*protobuf.*@-DPROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
+build --per_file_copt=external/.*grpc.*@-DGRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
diff --git a/MODULE.bazel b/MODULE.bazel
index 40b9552..9082a06 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -12,32 +12,32 @@ bazel_dep(name = "rules_pkg", version = "1.2.0", dev_dependency = True)
 bazel_dep(name = "container_structure_test", version = "1.21.1", dev_dependency = True)
 bazel_dep(name = "aspect_rules_lint", version = "1.10.2", dev_dependency = True)
 bazel_dep(name = "buildifier_prebuilt", version = "8.2.1", dev_dependency = True)
-bazel_dep(name = "rules", version = "0.20.37", dev_dependency = True)
+bazel_dep(name = "rules", version = "0.20.67", dev_dependency = True)
 git_override(
     module_name = "rules",
     remote = "git@github.com:mindful-hq/rules.git",
-    tag = "v0.20.37",
+    tag = "v0.20.67",
 )
 
-bazel_dep(name = "global_generics", version = "0.34.1", dev_dependency = True)
+bazel_dep(name = "global_generics", version = "0.38.11", dev_dependency = True)
 git_override(
     module_name = "global_generics",
     remote = "git@github.com:mindful-hq/global-generics.git",
-    tag = "v0.34.1",
+    tag = "v0.38.11",
 )
 
-bazel_dep(name = "global_proto", version = "2.17.4", dev_dependency = True)
+bazel_dep(name = "global_proto", version = "2.21.0", dev_dependency = True)
 git_override(
     module_name = "global_proto",
     remote = "git@github.com:mindful-hq/global-proto.git",
-    tag = "v2.17.4",
+    tag = "v2.21.0",
 )
 
-bazel_dep(name = "earth_resourcemanager_service_proto", version = "0.7.1", dev_dependency = True)
+bazel_dep(name = "earth_resourcemanager_service_proto", version = "0.7.2", dev_dependency = True)
 git_override(
     module_name = "earth_resourcemanager_service_proto",
     remote = "git@github.com:mindful-hq/earth-resourcemanager-service-proto.git",
-    tag = "v0.7.1",
+    tag = "v0.7.2",
 )
 
 bazel_dep(name = "earth_language_service_proto", version = "0.4.0", dev_dependency = True)
@@ -47,11 +47,11 @@ git_override(
     tag = "v0.4.0",
 )
 
-bazel_dep(name = "earth_user_service_proto", version = "0.22.2", dev_dependency = True)
+bazel_dep(name = "earth_user_service_proto", version = "0.22.3", dev_dependency = True)
 git_override(
     module_name = "earth_user_service_proto",
     remote = "git@github.com:mindful-hq/earth-user-service-proto.git",
-    tag = "v0.22.2",
+    tag = "v0.22.3",
 )
 
 bazel_dep(name = "earth_email_service_proto", version = "1.6.4", dev_dependency = True)
@@ -75,11 +75,11 @@ git_override(
     tag = "v0.4.0",
 )
 
-bazel_dep(name = "earth_authorization_service_proto", version = "0.5.2", dev_dependency = True)
+bazel_dep(name = "earth_authorization_service_proto", version = "0.6.0", dev_dependency = True)
 git_override(
     module_name = "earth_authorization_service_proto",
     remote = "git@github.com:mindful-hq/earth-authorization-service-proto.git",
-    tag = "v0.5.2",
+    tag = "v0.6.0",
 )
 
 ####################################################################
@@ -87,7 +87,7 @@ git_override(
 ####################################################################
 
 go_sdk = use_extension("@rules_go//go:extensions.bzl", "go_sdk")
-go_sdk.download(version = "1.24.1")
+go_sdk.download(version = "1.26.2")
 go_sdk.nogo(nogo = "//:nogo")
 
 ####################################################################
@@ -104,10 +104,16 @@ use_repo(
     "com_github_go_sql_driver_mysql",
     "com_github_google_uuid",
     "com_github_grpc_ecosystem_go_grpc_middleware_v2",
+    "com_github_pressly_goose_v3",
     "com_google_cloud_go_longrunning",
     "com_google_cloud_go_pubsub",
+    "io_ariga_atlas_provider_gorm",
     "io_gorm_driver_mysql",
     "io_gorm_gorm",
+    "io_opentelemetry_go_contrib_instrumentation_google_golang_org_grpc_otelgrpc",
+    "io_opentelemetry_go_contrib_propagators_b3",
+    "io_opentelemetry_go_otel",
+    "io_opentelemetry_go_otel_trace",
     "org_golang_google_genproto_googleapis_rpc",
     "org_golang_google_grpc",
     "org_golang_google_protobuf",
@@ -121,7 +127,7 @@ use_repo(
 oci = use_extension("@rules_oci//oci:extensions.bzl", "oci")
 oci.pull(
     name = "distroless_static",
-    digest = "sha256:6c8e62544dfb33087ab275c463747c75a5bb634a383af5e91625da882a5fe175",  # nonroot
+    digest = "sha256:963fa6c544fe5ce420f1f54fb88b6fb01479f054c8056d0f74cc2c6000df5240",  # nonroot
     image = "gcr.io/distroless/static",
     platforms = [
         "linux/amd64",
diff --git a/cmd/atlas/BUILD.bazel b/cmd/atlas/BUILD.bazel
new file mode 100644
index 0000000..2cef829
--- /dev/null
+++ b/cmd/atlas/BUILD.bazel
@@ -0,0 +1,30 @@
+load("@aspect_bazel_lib//lib:run_binary.bzl", "run_binary")
+load("@rules_go//go:def.bzl", "go_binary", "go_library")
+
+go_library(
+    name = "atlas_lib",
+    srcs = ["main.go"],
+    importpath = "github.com/mindful-hq/earth-email-service/cmd/atlas",
+    visibility = ["//visibility:private"],
+    deps = [
+        "//internal/database/model/model_v1",
+        "@io_ariga_atlas_provider_gorm//gormschema",
+    ],
+)
+
+go_binary(
+    name = "atlas",
+    embed = [":atlas_lib"],
+    visibility = ["//visibility:public"],
+)
+
+run_binary(
+    name = "atlas_schema",
+    outs = ["schema.sql"],
+    args = [
+        "$(location schema.sql)",
+    ],
+    tool = ":atlas",
+    visibility = ["//visibility:public"],
+)
diff --git a/cmd/atlas/main.go b/cmd/atlas/main.go
new file mode 100644
index 0000000..0a0033e
--- /dev/null
+++ b/cmd/atlas/main.go
@@ -0,0 +1,37 @@
+package main
+
+import (
+	"log"
+	"os"
+
+	"ariga.io/atlas-provider-gorm/gormschema"
+	modelv1 "github.com/mindful-hq/earth-email-service/internal/database/model/model_v1"
+)
+
+func main() {
+	var err error
+
+	if len(os.Args) != 2 {
+		log.Fatal("usage: atlas <output_file>")
+	}
+
+	var stmts string
+	if stmts, err = gormschema.New("mysql").Load(
+		new(modelv1.EmailConfig),
+		new(modelv1.EmailTemplate),
+		new(modelv1.Email),
+		new(modelv1.EmailSendEmailOutbox),
+		new(modelv1.OperationSubscription),
+		new(modelv1.Operation),
+	); err != nil {
+		log.Fatal(err)
+	}
+
+	if err = os.WriteFile(
+		os.Args[1],
+		[]byte(stmts),
+		0o644,
+	); err != nil {
+		log.Fatal(err)
+	}
+}
diff --git a/cmd/earth_email_service/BUILD.bazel b/cmd/earth_email_service/BUILD.bazel
index b3b75e0..0dceb3e 100644
--- a/cmd/earth_email_service/BUILD.bazel
+++ b/cmd/earth_email_service/BUILD.bazel
@@ -21,7 +21,7 @@ go_library(
     importpath = "github.com/mindful-hq/earth-email-service/cmd/earth_email_service",
     visibility = ["//visibility:private"],
     deps = [
-        "//internal/database/model/model_v1",
+        "//db",
         "//internal/database/seeder/email/config/config_v1",
         "//internal/database/seeder/email/email_v1",
         "//internal/database/seeder/email/template/template_v1",
@@ -43,6 +43,7 @@ go_library(
         "@com_github_grpc_ecosystem_go_grpc_middleware_v2//interceptors/recovery",
         "@com_github_grpc_ecosystem_go_grpc_middleware_v2//interceptors/selector",
         "@com_github_grpc_ecosystem_go_grpc_middleware_v2//metadata",
+        "@com_github_pressly_goose_v3//:goose",
         "@com_google_cloud_go_pubsub//:pubsub",
         "@earth_email_service_proto//mindful/earth/email/config/v1:config",
         "@earth_email_service_proto//mindful/earth/email/template/v1:template",
@@ -60,6 +61,11 @@ go_library(
         "@io_gorm_driver_mysql//:mysql",
         "@io_gorm_gorm//:gorm",
         "@io_gorm_gorm//logger",
+        "@io_opentelemetry_go_contrib_instrumentation_google_golang_org_grpc_otelgrpc//:otelgrpc",
+        "@io_opentelemetry_go_contrib_propagators_b3//:b3",
+        "@io_opentelemetry_go_otel//:otel",
+        "@io_opentelemetry_go_otel//propagation",
+        "@io_opentelemetry_go_otel_trace//:trace",
         "@org_golang_google_grpc//:grpc",
         "@org_golang_google_grpc//credentials/insecure",
         "@org_golang_google_grpc//health",
diff --git a/cmd/earth_email_service/main.go b/cmd/earth_email_service/main.go
index be0333c..6f6587f 100644
--- a/cmd/earth_email_service/main.go
+++ b/cmd/earth_email_service/main.go
@@ -23,7 +23,7 @@ import (
 	emailconfigprotov1 "github.com/mindful-hq/earth-email-service-proto/mindful/earth/email/config/v1"
 	emailtemplateprotov1 "github.com/mindful-hq/earth-email-service-proto/mindful/earth/email/template/v1"
 	emailprotov1 "github.com/mindful-hq/earth-email-service-proto/mindful/earth/email/v1"
-	modelv1 "github.com/mindful-hq/earth-email-service/internal/database/model/model_v1"
+	atlasdb "github.com/mindful-hq/earth-email-service/db"
 	emailconfigseederv1 "github.com/mindful-hq/earth-email-service/internal/database/seeder/email/config/config_v1"
 	emailseederv1 "github.com/mindful-hq/earth-email-service/internal/database/seeder/email/email_v1"
 	emailtemplateseederv1 "github.com/mindful-hq/earth-email-service/internal/database/seeder/email/template/template_v1"
@@ -48,6 +48,12 @@ import (
 	"github.com/mindful-hq/global-generics/pkg/logger"
 	loggerv1 "github.com/mindful-hq/global-generics/pkg/logger/logger_v1"
 	globallongrunningprotov1 "github.com/mindful-hq/global-proto/v2/mindful/global/longrunning/v1"
+	"github.com/pressly/goose/v3"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"go.opentelemetry.io/contrib/propagators/b3"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
 	"golang.org/x/sync/errgroup"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
@@ -62,6 +68,11 @@ import (
 func main() {
 	var err error
 
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		b3.New(),
+	))
+
 	var envLoggerDevelopment bool
 	if err = envsBool(map[*bool]string{
 		&envLoggerDevelopment: "LOGGER_DEVELOPMENT",
@@ -70,7 +81,9 @@ func main() {
 	}
 
 	var zapLogger *loggerv1.ZapLogger
-	if zapLogger, err = loggerv1.NewZapLogger(envLoggerDevelopment); err != nil {
+	if zapLogger, err = loggerv1.NewZapLogger(
+		envLoggerDevelopment,
+	); err != nil {
 		log.Fatal(err.Error())
 	}
 
@@ -130,7 +143,11 @@ func main() {
 	var db *gorm.DB
 	if db, err = dbConnection(
 		gormLogger,
-		mysql.Open(fmt.Sprintf("%s@tcp(127.0.0.1:3306)/%s?charset=utf8mb4&parseTime=True&loc=UTC", envDbUser, envDbDatabase)),
+		mysql.Open(fmt.Sprintf(
+			"%s@tcp(127.0.0.1:3306)/%s?charset=utf8mb4&parseTime=True&loc=UTC",
+			envDbUser,
+			envDbDatabase,
+		)),
 		time.Second*30,
 	); err != nil {
 		zapLogger.Fatal(err.Error())
@@ -149,13 +166,22 @@ func main() {
 	sqlDb.SetMaxOpenConns(envDbMaxOpenConnections)
 	sqlDb.SetConnMaxLifetime(time.Hour)
 
-	if err = db.AutoMigrate(
-		new(modelv1.EmailConfig),
-		new(modelv1.EmailTemplate),
-		new(modelv1.Email),
-		new(modelv1.EmailSendEmailOutbox),
-		new(modelv1.OperationSubscription),
-		new(modelv1.Operation),
+	ctx, ctxCancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer ctxCancel()
+
+	goose.SetBaseFS(atlasdb.Migrations)
+	goose.SetLogger(loggerv1.NewGooseLogger(zapLogger))
+
+	if err = goose.SetDialect(
+		"mysql",
+	); err != nil {
+		zapLogger.Fatal(err.Error())
+	}
+
+	if err = goose.UpContext(
+		ctx,
+		sqlDb,
+		"atlas_migrations_out",
 	); err != nil {
 		zapLogger.Fatal(err.Error())
 	}
@@ -219,9 +245,6 @@ func main() {
 		earthLanguageServiceConn,
 	)
 
-	ctx, ctxCancel := context.WithTimeout(context.Background(), 10*time.Second)
-	defer ctxCancel()
-
 	var googlePubSubClient *googlepubsub.Client
 	if googlePubSubClient, err = googlepubsub.NewClient(
 		ctx,
@@ -277,6 +300,7 @@ func main() {
 	}
 
 	grpcServer := grpc.NewServer(
+		grpc.StatsHandler(otelgrpc.NewServerHandler()),
 		grpc.ChainUnaryInterceptor(
 			selector.UnaryServerInterceptor(
 				logging.UnaryServerInterceptor(interceptorLogger, grpcLoggerOpts...),
@@ -336,7 +360,7 @@ func main() {
 		}
 	}()
 
-	zapLogger.Info(fmt.Sprintf("grpc papa1234 server started on port %s", envGrpcPort))
+	zapLogger.Info(fmt.Sprintf("grpc server started on port %s", envGrpcPort))
 
 	for _, pubSubSubscription := range []interface {
 		GetSubscriptionId() string
@@ -392,7 +416,7 @@ func main() {
 			defer func() {
 				if r := recover(); r != nil {
 					zapLogger.Error(fmt.Errorf("%v", r).Error())
-					err = statusError.InternalError().Err()
+					err = fmt.Errorf("%v", r)
 				}
 			}()
 
@@ -411,12 +435,11 @@ func main() {
 	for _, outbox := range []outbox.Outbox{
 		emailSendEmailOutbox,
 	} {
-		outbox := outbox // https://github.com/bazel-contrib/rules_go/issues/3924
 		groupOutboxes.Go(func() (err error) {
 			defer func() {
 				if r := recover(); r != nil {
 					zapLogger.Error(fmt.Errorf("%v", r).Error())
-					err = statusError.InternalError().Err()
+					err = fmt.Errorf("%v", r)
 				}
 			}()
 
@@ -473,7 +496,7 @@ func envsInt(envIndex map[*int]string) error {
 			return fmt.Errorf("env not exists: %s", envName)
 		}
 
-		if *env, err = strconv.Atoi(envTmp); !ok {
+		if *env, err = strconv.Atoi(envTmp); err != nil {
 			return err
 		}
 	}
@@ -490,7 +513,7 @@ func envsBool(envIndex map[*bool]string) error {
 			return fmt.Errorf("env not exists: %s", envName)
 		}
 
-		if *env, err = strconv.ParseBool(envTmp); !ok {
+		if *env, err = strconv.ParseBool(envTmp); err != nil {
 			return err
 		}
 	}
@@ -525,15 +548,20 @@ func dbConnection(gormLogger gormlogger.Interface, dialector gorm.Dialector, tim
 }
 
 func loggerFieldsFromContext(ctx context.Context) logger.Fields {
-	md := metadata.ExtractIncoming(ctx)
-	fieldIndex := map[string]string{
+	traceIdLogKey := "grpc.trace_id"
+	mdIndex := map[string]string{
 		"x-mindful-email": "grpc.email",
 		"x-request-id":    "grpc.request_id",
-		"x-b3-traceid":    "grpc.trace_id",
 	}
 
-	fields := make(logger.Fields, 0, len(fieldIndex)*2)
-	for mdKey, logKey := range fieldIndex {
+	fields := make(logger.Fields, 0, 2+len(mdIndex)*2)
+
+	if spanCtx := trace.SpanContextFromContext(ctx); spanCtx.HasTraceID() {
+		fields = append(fields, traceIdLogKey, spanCtx.TraceID().String())
+	}
+
+	md := metadata.ExtractIncoming(ctx)
+	for mdKey, logKey := range mdIndex {
 		if mdValue := md.Get(mdKey); mdValue != "" {
 			fields = append(fields, logKey, mdValue)
 		}
diff --git a/cmd/earth_emailauthorization_service/BUILD.bazel b/cmd/earth_emailauthorization_service/BUILD.bazel
index 6ff2811..6ab2ce1 100644
--- a/cmd/earth_emailauthorization_service/BUILD.bazel
+++ b/cmd/earth_emailauthorization_service/BUILD.bazel
@@ -32,6 +32,11 @@ go_library(
         "@global_generics//pkg/grpc/status_error/status_error_v1",
         "@global_generics//pkg/logger",
         "@global_generics//pkg/logger/logger_v1",
+        "@io_opentelemetry_go_contrib_instrumentation_google_golang_org_grpc_otelgrpc//:otelgrpc",
+        "@io_opentelemetry_go_contrib_propagators_b3//:b3",
+        "@io_opentelemetry_go_otel//:otel",
+        "@io_opentelemetry_go_otel//propagation",
+        "@io_opentelemetry_go_otel_trace//:trace",
         "@org_golang_google_grpc//:grpc",
         "@org_golang_google_grpc//credentials/insecure",
         "@org_golang_google_grpc//health",
diff --git a/cmd/earth_emailauthorization_service/main.go b/cmd/earth_emailauthorization_service/main.go
index d75d82a..e5cc845 100644
--- a/cmd/earth_emailauthorization_service/main.go
+++ b/cmd/earth_emailauthorization_service/main.go
@@ -25,6 +25,11 @@ import (
 	statuserrorv1 "github.com/mindful-hq/global-generics/pkg/grpc/status_error/status_error_v1"
 	logger "github.com/mindful-hq/global-generics/pkg/logger"
 	loggerv1 "github.com/mindful-hq/global-generics/pkg/logger/logger_v1"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"go.opentelemetry.io/contrib/propagators/b3"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
 	"google.golang.org/grpc/health"
@@ -34,6 +39,11 @@ import (
 func main() {
 	var err error
 
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		b3.New(),
+	))
+
 	var envLoggerDevelopment bool
 	if err = envsBool(map[*bool]string{
 		&envLoggerDevelopment: "LOGGER_DEVELOPMENT",
@@ -42,7 +52,9 @@ func main() {
 	}
 
 	var zapLogger *loggerv1.ZapLogger
-	if zapLogger, err = loggerv1.NewZapLogger(envLoggerDevelopment); err != nil {
+	if zapLogger, err = loggerv1.NewZapLogger(
+		envLoggerDevelopment,
+	); err != nil {
 		log.Fatal(err.Error())
 	}
 
@@ -96,6 +108,7 @@ func main() {
 	}
 
 	grpcServer := grpc.NewServer(
+		grpc.StatsHandler(otelgrpc.NewServerHandler()),
 		grpc.ChainUnaryInterceptor(
 			selector.UnaryServerInterceptor(
 				logging.UnaryServerInterceptor(interceptorLogger, grpcLoggerOpts...),
@@ -201,7 +214,7 @@ func envsBool(envIndex map[*bool]string) error {
 			return fmt.Errorf("env not exists: %s", envName)
 		}
 
-		if *env, err = strconv.ParseBool(envTmp); !ok {
+		if *env, err = strconv.ParseBool(envTmp); err != nil {
 			return err
 		}
 	}
@@ -210,13 +223,20 @@ func envsBool(envIndex map[*bool]string) error {
 }
 
 func loggerFieldsFromContext(ctx context.Context) logger.Fields {
-	md := metadata.ExtractIncoming(ctx)
-	fieldIndex := map[string]string{
-		"x-b3-traceid": "grpc.trace_id",
+	traceIdLogKey := "grpc.trace_id"
+	mdIndex := map[string]string{
+		"x-mindful-email": "grpc.email",
+		"x-request-id":    "grpc.request_id",
 	}
 
-	fields := make(logger.Fields, 0, len(fieldIndex)*2)
-	for mdKey, logKey := range fieldIndex {
+	fields := make(logger.Fields, 0, 2+len(mdIndex)*2)
+
+	if spanCtx := trace.SpanContextFromContext(ctx); spanCtx.HasTraceID() {
+		fields = append(fields, traceIdLogKey, spanCtx.TraceID().String())
+	}
+
+	md := metadata.ExtractIncoming(ctx)
+	for mdKey, logKey := range mdIndex {
 		if mdValue := md.Get(mdKey); mdValue != "" {
 			fields = append(fields, logKey, mdValue)
 		}
diff --git a/db/BUILD.bazel b/db/BUILD.bazel
new file mode 100644
index 0000000..d1f1111
--- /dev/null
+++ b/db/BUILD.bazel
@@ -0,0 +1,38 @@
+load("@aspect_bazel_lib//lib:write_source_files.bzl", "write_source_files")
+load("@rules//pkg/atlas:atlas.bzl", "atlas_migrations_mysql", "atlas_migrations_mysql_test")
+load("@rules_go//go:def.bzl", "go_library")
+
+atlas_migrations_mysql(
+    name = "atlas_migrations",
+    migrations = glob(
+        [
+            "migrations/*.sql",
+            "migrations/atlas.sum",
+        ],
+        allow_empty = True,
+    ),
+    schema = "//cmd/atlas:atlas_schema",
+)
+
+# gazelle:exclude db.go
+# keep
+go_library(
+    name = "db",
+    srcs = ["db.go"],
+    embedsrcs = [":atlas_migrations"],
+    importpath = "github.com/mindful-hq/earth-email-service/db",
+    visibility = ["//cmd/earth_email_service:__pkg__"],
+)
+
+atlas_migrations_mysql_test(
+    name = "atlas_migrations_test",
+    migrations = ":atlas_migrations",
+)
+
+write_source_files(
+    name = "atlas_migration_sources",
+    files = {
+        "migrations": ":atlas_migrations",
+    },
+    visibility = ["//visibility:public"],
+)
diff --git a/db/db.go b/db/db.go
new file mode 100644
index 0000000..e304281
--- /dev/null
+++ b/db/db.go
@@ -0,0 +1,6 @@
+package db
+
+import "embed"
+
+//go:embed atlas_migrations_out/*.sql
+var Migrations embed.FS
diff --git a/go.mod b/go.mod
index d010d8c..b8289e7 100644
--- a/go.mod
+++ b/go.mod
@@ -1,74 +1,108 @@
 module github.com/mindful-hq/earth-email-service
 
-go 1.24.0
-
-toolchain go1.24.1
+go 1.26.2
 
 require (
+	ariga.io/atlas-provider-gorm v0.6.0
 	buf.build/go/protovalidate v1.0.0
 	cloud.google.com/go/longrunning v0.6.7
-	cloud.google.com/go/pubsub v1.49.0
-	github.com/envoyproxy/go-control-plane/envoy v1.32.4
-	github.com/go-sql-driver/mysql v1.9.2
+	cloud.google.com/go/pubsub v1.50.0
+	github.com/envoyproxy/go-control-plane/envoy v1.36.0
+	github.com/go-sql-driver/mysql v1.9.3
 	github.com/google/uuid v1.6.0
 	github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.3.2
-	github.com/mindful-hq/earth-authorization-service-proto v0.5.2
+	github.com/mindful-hq/earth-authorization-service-proto v0.6.0
 	github.com/mindful-hq/earth-email-service-internal-proto v0.4.2
 	github.com/mindful-hq/earth-email-service-proto v1.6.4
 	github.com/mindful-hq/earth-emailmailgun-service-internal-proto v0.4.0
 	github.com/mindful-hq/earth-language-service-proto v0.4.0
-	github.com/mindful-hq/earth-resourcemanager-service-proto v0.7.1
-	github.com/mindful-hq/earth-user-service-proto v0.22.2
-	github.com/mindful-hq/global-generics v0.34.1
-	github.com/mindful-hq/global-proto/v2 v2.17.5
-	golang.org/x/sync v0.17.0
-	google.golang.org/genproto/googleapis/rpc v0.0.0-20250804133106-a7a43d27e69b
-	google.golang.org/grpc v1.74.2
-	google.golang.org/protobuf v1.36.9
-	gorm.io/driver/mysql v1.5.7
-	gorm.io/gorm v1.30.0
+	github.com/mindful-hq/earth-resourcemanager-service-proto v0.7.2
+	github.com/mindful-hq/earth-user-service-proto v0.22.3
+	github.com/mindful-hq/global-generics v0.38.11
+	github.com/mindful-hq/global-proto/v2 v2.21.0
+	github.com/pressly/goose/v3 v3.27.1
+	go.opentelemetry.io/contrib/propagators/b3 v1.43.0
+	golang.org/x/sync v0.20.0
+	google.golang.org/genproto/googleapis/rpc v0.0.0-20260420184626-e10c466a9529
+	google.golang.org/grpc v1.80.0
+	google.golang.org/protobuf v1.36.11
+	gorm.io/driver/mysql v1.6.0
+	gorm.io/gorm v1.31.1
+)
+
+require (
+	ariga.io/atlas v0.36.2-0.20250806044935-5bb51a0a956e // indirect
+	cloud.google.com/go/monitoring v1.24.2 // indirect
+	cloud.google.com/go/pubsub/v2 v2.0.0 // indirect
+	cloud.google.com/go/spanner v1.84.1 // indirect
+	github.com/GoogleCloudPlatform/grpc-gcp-go/grpcgcp v1.5.3 // indirect
+	github.com/GoogleCloudPlatform/opentelemetry-operations-go/detectors/gcp v1.31.0 // indirect
+	github.com/cespare/xxhash/v2 v2.3.0 // indirect
+	github.com/go-jose/go-jose/v4 v4.1.3 // indirect
+	github.com/golang-sql/civil v0.0.0-20220223132316-b832511892a9 // indirect
+	github.com/golang-sql/sqlexp v0.1.0 // indirect
+	github.com/golang/groupcache v0.0.0-20241129210726-2c02b8208cf8 // indirect
+	github.com/googleapis/go-gorm-spanner v1.8.6 // indirect
+	github.com/googleapis/go-sql-spanner v1.17.0 // indirect
+	github.com/hashicorp/golang-lru/v2 v2.0.7 // indirect
+	github.com/jackc/pgpassfile v1.0.0 // indirect
+	github.com/jackc/pgservicefile v0.0.0-20240606120523-5a60cdf6a761 // indirect
+	github.com/jackc/pgx/v5 v5.9.2 // indirect
+	github.com/jackc/puddle/v2 v2.2.2 // indirect
+	github.com/mattn/go-sqlite3 v1.14.28 // indirect
+	github.com/mfridman/interpolate v0.0.2 // indirect
+	github.com/microsoft/go-mssqldb v1.9.8 // indirect
+	github.com/sethvargo/go-retry v0.3.0 // indirect
+	github.com/shopspring/decimal v1.4.0 // indirect
+	github.com/spiffe/go-spiffe/v2 v2.6.0 // indirect
+	go.opentelemetry.io/contrib/detectors/gcp v1.39.0 // indirect
+	go.opentelemetry.io/otel/sdk v1.43.0 // indirect
+	go.opentelemetry.io/otel/sdk/metric v1.43.0 // indirect
+	gorm.io/driver/postgres v1.5.11 // indirect
+	gorm.io/driver/sqlite v1.6.0 // indirect
+	gorm.io/driver/sqlserver v1.5.4 // indirect
 )
 
 require (
-	buf.build/gen/go/bufbuild/protovalidate/protocolbuffers/go v1.36.9-20250912141014-52f32327d4b0.1 // indirect
-	cel.dev/expr v0.24.0 // indirect
-	cloud.google.com/go v0.120.0 // indirect
-	cloud.google.com/go/auth v0.16.0 // indirect
+	buf.build/gen/go/bufbuild/protovalidate/protocolbuffers/go v1.36.11-20260209202127-80ab13bee0bf.1 // indirect
+	cel.dev/expr v0.25.1 // indirect
+	cloud.google.com/go v0.121.6 // indirect
+	cloud.google.com/go/auth v0.16.4 // indirect
 	cloud.google.com/go/auth/oauth2adapt v0.2.8 // indirect
-	cloud.google.com/go/compute/metadata v0.7.0 // indirect
+	cloud.google.com/go/compute/metadata v0.9.0 // indirect
 	cloud.google.com/go/iam v1.5.2 // indirect
-	filippo.io/edwards25519 v1.1.0 // indirect
+	filippo.io/edwards25519 v1.2.0 // indirect
 	github.com/antlr4-go/antlr/v4 v4.13.1 // indirect
-	github.com/cncf/xds/go v0.0.0-20250501225837-2ac532fd4443 // indirect
-	github.com/envoyproxy/protoc-gen-validate v1.2.1 // indirect
+	github.com/cncf/xds/go v0.0.0-20251210132809-ee656c7534f5 // indirect
+	github.com/envoyproxy/protoc-gen-validate v1.3.0 // indirect
 	github.com/felixge/httpsnoop v1.0.4 // indirect
 	github.com/go-logr/logr v1.4.3 // indirect
 	github.com/go-logr/stdr v1.2.2 // indirect
 	github.com/google/cel-go v0.26.1 // indirect
 	github.com/google/s2a-go v0.1.9 // indirect
 	github.com/googleapis/enterprise-certificate-proxy v0.3.6 // indirect
-	github.com/googleapis/gax-go/v2 v2.14.1 // indirect
+	github.com/googleapis/gax-go/v2 v2.15.0 // indirect
 	github.com/jinzhu/inflection v1.0.0 // indirect
 	github.com/jinzhu/now v1.1.5 // indirect
 	github.com/planetscale/vtprotobuf v0.6.1-0.20240319094008-0393e58bdf10 // indirect
 	github.com/stoewer/go-strcase v1.3.1 // indirect
 	go.opencensus.io v0.24.0 // indirect
-	go.opentelemetry.io/auto/sdk v1.1.0 // indirect
-	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc v0.60.0 // indirect
-	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.60.0 // indirect
-	go.opentelemetry.io/otel v1.36.0 // indirect
-	go.opentelemetry.io/otel/metric v1.36.0 // indirect
-	go.opentelemetry.io/otel/trace v1.36.0 // indirect
+	go.opentelemetry.io/auto/sdk v1.2.1 // indirect
+	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc v0.67.0
+	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.68.0 // indirect
+	go.opentelemetry.io/otel v1.43.0
+	go.opentelemetry.io/otel/metric v1.43.0 // indirect
+	go.opentelemetry.io/otel/trace v1.43.0
 	go.uber.org/multierr v1.11.0 // indirect
 	go.uber.org/zap v1.27.0 // indirect
-	golang.org/x/crypto v0.40.0 // indirect
-	golang.org/x/exp v0.0.0-20250620022241-b7579e27df2b // indirect
-	golang.org/x/net v0.42.0 // indirect
-	golang.org/x/oauth2 v0.30.0 // indirect
-	golang.org/x/sys v0.34.0 // indirect
-	golang.org/x/text v0.27.0 // indirect
-	golang.org/x/time v0.11.0 // indirect
-	google.golang.org/api v0.229.0 // indirect
-	google.golang.org/genproto v0.0.0-20250715232539-7130f93afb79 // indirect
-	google.golang.org/genproto/googleapis/api v0.0.0-20250804133106-a7a43d27e69b // indirect
+	golang.org/x/crypto v0.50.0 // indirect
+	golang.org/x/exp v0.0.0-20260410095643-746e56fc9e2f // indirect
+	golang.org/x/net v0.53.0 // indirect
+	golang.org/x/oauth2 v0.34.0 // indirect
+	golang.org/x/sys v0.43.0 // indirect
+	golang.org/x/text v0.36.0 // indirect
+	golang.org/x/time v0.12.0 // indirect
+	google.golang.org/api v0.247.0 // indirect
+	google.golang.org/genproto v0.0.0-20250804133106-a7a43d27e69b // indirect
+	google.golang.org/genproto/googleapis/api v0.0.0-20260401024825-9d38bb4040a9 // indirect
 )
diff --git a/cmd/earth_email_service/main.go b/cmd/earth_email_service/main.go
index 20c5e3a..c7c3969 100644
--- a/cmd/earth_auth_service/main.go
+++ b/cmd/earth_auth_service/main.go
@@ -213,7 +213,7 @@ func main() {
 	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
 	<-quit
 
-	zapLogger.Info(fmt.Sprintf("shutting down server(s) ..."))
+	zapLogger.Info("shutting down server(s) ...")
 
 	grpcServer.GracefulStop()
 	zapLogger.Info("grpc server: shutdown successfully")
